Loading a mergedGrid does not work in current MoBIE-beta

Loading a dataset that contains a mergedGrid fails with:
```
java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.RuntimeException: Module threw exception
	at net.imagej.legacy.LegacyService.runLegacyCompatibleCommand(LegacyService.java:307)
	at net.imagej.legacy.DefaultLegacyHooks.interceptRunPlugIn(DefaultLegacyHooks.java:166)
	at ij.IJ.runPlugIn(IJ.java)
	at ij.Executer.runCommand(Executer.java:150)
	at ij.Executer.run(Executer.java:68)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.util.concurrent.ExecutionException: java.lang.RuntimeException: Module threw exception
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at net.imagej.legacy.LegacyService.runLegacyCompatibleCommand(LegacyService.java:303)
	... 5 more
Caused by: java.lang.RuntimeException: Module threw exception
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:127)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:63)
	at org.scijava.thread.DefaultThreadService.lambda$wrap$2(DefaultThreadService.java:225)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	... 1 more
Caused by: java.lang.RuntimeException: Unknown class: mergedGrid
	at org.embl.mobie.viewer.serialize.JsonHelper.createObjectFromJsonValue(JsonHelper.java:45)
	at org.embl.mobie.viewer.serialize.SourceTransformerListAdapter.deserialize(SourceTransformerListAdapter.java:38)
	at org.embl.mobie.viewer.serialize.SourceTransformerListAdapter.deserialize(SourceTransformerListAdapter.java:14)
	at com.google.gson.internal.bind.TreeTypeAdapter.read(TreeTypeAdapter.java:69)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:131)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:222)
	at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:41)
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:187)
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:145)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:131)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:222)
	at com.google.gson.Gson.fromJson(Gson.java:932)
	at com.google.gson.Gson.fromJson(Gson.java:897)
	at com.google.gson.Gson.fromJson(Gson.java:846)
	at org.embl.mobie.viewer.serialize.DatasetJsonParser.parseDataset(DatasetJsonParser.java:24)
	at org.embl.mobie.viewer.MoBIE.openDataset(MoBIE.java:220)
	at org.embl.mobie.viewer.MoBIE.openDataset(MoBIE.java:160)
	at org.embl.mobie.viewer.MoBIE.<init>(MoBIE.java:96)
	at org.embl.mobie.viewer.command.OpenMoBIEProjectCommand.run(OpenMoBIEProjectCommand.java:25)
	at org.scijava.command.CommandModule.run(CommandModule.java:196)
	at org.scijava.module.ModuleRunner.run(ModuleRunner.java:163)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:124)
	... 6 more
```

To reproduce: open the project at /g/kreshuk/pape/Work/mobie/covid-if-project (data is not uploaded to s3 yet)

Now there's a different error:
```
java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.RuntimeException: Module threw exception
	at net.imagej.legacy.LegacyService.runLegacyCompatibleCommand(LegacyService.java:307)
	at net.imagej.legacy.DefaultLegacyHooks.interceptRunPlugIn(DefaultLegacyHooks.java:166)
	at ij.IJ.runPlugIn(IJ.java)
	at ij.Executer.runCommand(Executer.java:150)
	at ij.Executer.run(Executer.java:68)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.util.concurrent.ExecutionException: java.lang.RuntimeException: Module threw exception
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at net.imagej.legacy.LegacyService.runLegacyCompatibleCommand(LegacyService.java:303)
	... 5 more
Caused by: java.lang.RuntimeException: Module threw exception
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:127)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:63)
	at org.scijava.thread.DefaultThreadService.lambda$wrap$2(DefaultThreadService.java:225)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	... 1 more
Caused by: java.lang.NullPointerException
	at org.embl.mobie.viewer.MoBIE.getSource(MoBIE.java:289)
	at org.embl.mobie.viewer.MoBIE.loadImageSegmentsTable(MoBIE.java:366)
	at org.embl.mobie.viewer.MoBIE.loadPrimarySegmentsTables(MoBIE.java:414)
	at org.embl.mobie.viewer.MoBIE.loadPrimarySegmentsTables(MoBIE.java:487)
	at org.embl.mobie.viewer.view.ViewManager.loadTablesAndCreateImageSegments(ViewManager.java:452)
	at org.embl.mobie.viewer.view.ViewManager.showSegmentationDisplay(ViewManager.java:407)
	at org.embl.mobie.viewer.view.ViewManager.showSourceDisplay(ViewManager.java:338)
	at org.embl.mobie.viewer.view.ViewManager.show(ViewManager.java:289)
	at org.embl.mobie.viewer.MoBIE.openDataset(MoBIE.java:215)
	at org.embl.mobie.viewer.MoBIE.openDataset(MoBIE.java:149)
	at org.embl.mobie.viewer.MoBIE.<init>(MoBIE.java:85)
	at org.embl.mobie.viewer.command.OpenMoBIEProjectCommand.run(OpenMoBIEProjectCommand.java:25)
	at org.scijava.command.CommandModule.run(CommandModule.java:196)
	at org.scijava.module.ModuleRunner.run(ModuleRunner.java:163)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:124)
	... 6 more
```

--- a/src/main/java/org/embl/mobie/viewer/MoBIE.java
+++ b/src/main/java/org/embl/mobie/viewer/MoBIE.java
@@ -278,18 +278,9 @@ public class MoBIE
 //		userInterface.dispose();
 	}
 
-	public synchronized ImageSource getSource(String sourceName )
+	public synchronized ImageSource getSource( String sourceName )
 	{
-		try
-		{
-			return dataset.sources.get( sourceName ).get();
-		}
-		catch ( Exception e )
-		{
-			System.err.println( "Could not find source " + sourceName + " among the available sources in this dataset.");
-			e.printStackTrace();
-			throw new RuntimeException();
-		}
+		return dataset.sources.get( sourceName ).get();
 	}
 
 	public SourceAndConverter< ? > openSourceAndConverter( String sourceName )

--- a/src/main/java/org/embl/mobie/viewer/MoBIE.java
+++ b/src/main/java/org/embl/mobie/viewer/MoBIE.java
@@ -404,7 +404,7 @@ public class MoBIE
 
 		for ( String sourceName : segmentationDisplay.getSources() )
 		{
-			fetchRootSources( getSourceAndConverter( sourceName ).getSpimSource(), rootSourceNames );
+			fetchTableRootSources( getSourceAndConverter( sourceName ).getSpimSource(), rootSourceNames );
 		}
 
 		// TODO: make parallel
@@ -422,7 +422,7 @@ public class MoBIE
 	 * @param source
 	 * @param rootSourceNames
 	 */
-	private void fetchRootSources( Source< ? > source, Set< String > rootSourceNames )
+	private void fetchTableRootSources( Source< ? > source, Set< String > rootSourceNames )
 	{
 		if ( source instanceof SpimSource )
 		{
@@ -432,7 +432,7 @@ public class MoBIE
 		{
 			final Source< ? > wrappedSource = ( ( TransformedSource ) source ).getWrappedSource();
 
-			fetchRootSources( wrappedSource, rootSourceNames );
+			fetchTableRootSources( wrappedSource, rootSourceNames );
 		}
 		else if (  source instanceof MergedGridSource )
 		{
@@ -440,8 +440,12 @@ public class MoBIE
 			final List< ? extends Source< ? > > gridSources = mergedGridSource.getGridSources();
 			for ( Source< ? > gridSource : gridSources )
 			{
-				fetchRootSources( gridSource, rootSourceNames );
+				fetchTableRootSources( gridSource, rootSourceNames );
 			}
+		}
+		else
+		{
+			throw new IllegalArgumentException("Sources of type " + source.getClass().getName() + " are currently not supported in MoBIE.fetchTableRootSources.");
 		}
 	}
 

--- a/src/main/java/org/embl/mobie/viewer/MoBIE.java
+++ b/src/main/java/org/embl/mobie/viewer/MoBIE.java
@@ -1,19 +1,17 @@
 package org.embl.mobie.viewer;
 
+import bdv.SpimSource;
 import bdv.tools.transformation.TransformedSource;
-import bdv.util.BdvFunctions;
 import bdv.util.volatiles.SharedQueue;
 import bdv.img.n5.N5ImageLoader;
 import bdv.viewer.Source;
 import bdv.viewer.SourceAndConverter;
 import de.embl.cba.bdv.utils.Logger;
-import net.imglib2.realtransform.AffineTransform3D;
 import org.embl.mobie.io.ome.zarr.loaders.N5OMEZarrImageLoader;
 import org.embl.mobie.viewer.display.SegmentationSourceDisplay;
 import org.embl.mobie.viewer.display.AnnotatedIntervalDisplay;
 import org.embl.mobie.viewer.annotate.AnnotatedIntervalCreator;
 import org.embl.mobie.viewer.annotate.AnnotatedIntervalTableRow;
-import org.embl.mobie.viewer.playground.SourceChanger;
 import org.embl.mobie.viewer.serialize.DatasetJsonParser;
 import org.embl.mobie.viewer.serialize.ProjectJsonParser;
 import org.embl.mobie.viewer.source.ImageDataFormat;
@@ -21,6 +19,7 @@ import org.embl.mobie.viewer.source.ImageSource;
 import org.embl.mobie.viewer.source.SegmentationSource;
 import org.embl.mobie.viewer.source.SpimDataOpener;
 import org.embl.mobie.viewer.table.TableDataFormat;
+import org.embl.mobie.viewer.transform.MergedGridSource;
 import org.embl.mobie.viewer.ui.UserInterface;
 import org.embl.mobie.viewer.ui.WindowArrangementHelper;
 import org.embl.mobie.viewer.view.View;
@@ -34,10 +33,8 @@ import ij.IJ;
 import mpicbg.spim.data.SpimData;
 import mpicbg.spim.data.sequence.ImgLoader;
 import sc.fiji.bdvpg.PlaygroundPrefs;
-import sc.fiji.bdvpg.scijava.services.SourceAndConverterService;
 import sc.fiji.bdvpg.services.SourceAndConverterServices;
 import sc.fiji.bdvpg.sourceandconverter.importer.SourceAndConverterFromSpimDataCreator;
-import sc.fiji.bdvpg.sourceandconverter.transform.SourceAffineTransformer;
 
 import java.io.IOException;
 import java.util.*;
@@ -45,7 +42,6 @@ import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
-import java.util.function.Function;
 
 public class MoBIE
 {
@@ -64,7 +60,6 @@ public class MoBIE
 	private String imageRoot;
 	private String tableRoot;
 	private HashMap< String, ImgLoader > sourceNameToImgLoader;
-	private SourceAndConverterService sacService;
 	private Map< String, SourceAndConverter< ? > > sourceNameToSourceAndConverter;
 
 	public MoBIE( String projectRoot ) throws IOException
@@ -81,7 +76,6 @@ public class MoBIE
 		PlaygroundPrefs.setSourceAndConverterUIVisibility( false );
 		project = new ProjectJsonParser().parseProject( FileAndUrlUtils.combinePath( projectRoot,  "project.json" ) );
 		this.settings = setImageDataFormat( projectLocation );
-		sacService = ( SourceAndConverterService ) SourceAndConverterServices.getSourceAndConverterService();
 		openDataset();
 	}
 
@@ -415,16 +409,55 @@ public class MoBIE
 
 	private ArrayList< List< TableRowImageSegment > > loadPrimarySegmentsTables(SegmentationSourceDisplay segmentationDisplay, String table )
 	{
-		final ArrayList< List< TableRowImageSegment > > primaryTables = new ArrayList<>();
+		final Set< String > rootSourceNames = ConcurrentHashMap.newKeySet();
 
-		// TODO: make parallel
 		for ( String sourceName : segmentationDisplay.getSources() )
 		{
-			final List< TableRowImageSegment > primaryTable = loadImageSegmentsTable( sourceName, table );
-			primaryTables.add( primaryTable );
+			fetchRootSources( getSourceAndConverter( sourceName ).getSpimSource(), rootSourceNames );
+		}
+
+		// TODO: make parallel
+		final ArrayList< List< TableRowImageSegment > > primaryTables = new ArrayList<>();
+		for ( String sourceName : rootSourceNames )
+		{
+			addPrimaryTable( table, primaryTables, sourceName );
 		}
 
 		return primaryTables;
+	}
+
+	/**
+	 * Recursively fetch all root sources
+	 * @param source
+	 * @param rootSourceNames
+	 */
+	private void fetchRootSources( Source< ? > source, Set< String > rootSourceNames )
+	{
+		if ( source instanceof SpimSource )
+		{
+			rootSourceNames.add( source.getName() );
+		}
+		else if ( source instanceof TransformedSource )
+		{
+			final Source< ? > wrappedSource = ( ( TransformedSource ) source ).getWrappedSource();
+
+			fetchRootSources( wrappedSource, rootSourceNames );
+		}
+		else if (  source instanceof MergedGridSource )
+		{
+			final MergedGridSource< ? > mergedGridSource = ( MergedGridSource ) source;
+			final List< ? extends Source< ? > > gridSources = mergedGridSource.getGridSources();
+			for ( Source< ? > gridSource : gridSources )
+			{
+				fetchRootSources( gridSource, rootSourceNames );
+			}
+		}
+	}
+
+	private void addPrimaryTable( String table, ArrayList< List< TableRowImageSegment > > primaryTables, String sourceName )
+	{
+		final List< TableRowImageSegment > primaryTable = loadImageSegmentsTable( sourceName, table );
+		primaryTables.add( primaryTable );
 	}
 
 	private Map< String, List< String > > createColumnsForMerging( List< TableRowImageSegment > segments, Map< String, List< String > > newColumns )

--- a/src/main/java/org/embl/mobie/viewer/transform/MergedGridSourceTransformer.java
+++ b/src/main/java/org/embl/mobie/viewer/transform/MergedGridSourceTransformer.java
@@ -28,7 +28,6 @@ public class MergedGridSourceTransformer extends AbstractSourceTransformer
 	private transient MergedGridSource< ? > mergedGridSource;
 	private transient double[] translationRealOffset;
 
-
 	@Override
 	public void transform( Map< String, SourceAndConverter< ? > > sourceNameToSourceAndConverter )
 	{
@@ -40,6 +39,11 @@ public class MergedGridSourceTransformer extends AbstractSourceTransformer
 		SourceAndConverter< ? > mergedSourceAndConverter = createMergedSourceAndConverter( gridSources.stream().map( sac -> sac.getSpimSource() ).collect( Collectors.toList() ), gridSources.get( 0 ).asVolatile().getConverter(), gridSources.get( 0 ).getConverter() );
 
 		sourceNameToSourceAndConverter.put( mergedSourceAndConverter.getSpimSource().getName(), mergedSourceAndConverter );
+
+		if ( mergedGridSourceName.equals( "plate_cell_segmentation" ) )
+		{
+			int a = 1;
+		}
 
 		// needed to know where the individual sources are in space:
 		transformGridSourcesIndividually( sourceNameToSourceAndConverter, gridSources );

--- a/src/main/java/org/embl/mobie/viewer/transform/MergedGridSourceTransformer.java
+++ b/src/main/java/org/embl/mobie/viewer/transform/MergedGridSourceTransformer.java
@@ -40,11 +40,6 @@ public class MergedGridSourceTransformer extends AbstractSourceTransformer
 
 		sourceNameToSourceAndConverter.put( mergedSourceAndConverter.getSpimSource().getName(), mergedSourceAndConverter );
 
-		if ( mergedGridSourceName.equals( "plate_cell_segmentation" ) )
-		{
-			int a = 1;
-		}
-
 		// needed to know where the individual sources are in space:
 		transformGridSourcesIndividually( sourceNameToSourceAndConverter, gridSources );
 	}

--- a/src/main/java/org/embl/mobie/viewer/view/ViewManager.java
+++ b/src/main/java/org/embl/mobie/viewer/view/ViewManager.java
@@ -4,6 +4,7 @@ import bdv.tools.transformation.TransformedSource;
 import bdv.util.BdvHandle;
 import bdv.viewer.Source;
 import bdv.viewer.SourceAndConverter;
+import de.embl.cba.tables.Logger;
 import de.embl.cba.tables.color.ColoringModel;
 import de.embl.cba.tables.color.ColumnColoringModelCreator;
 import de.embl.cba.tables.select.DefaultSelectionModel;
@@ -146,8 +147,15 @@ public class ViewManager
 		Map<String, String> sourceNameToTableDir = new HashMap<>();
 		for ( String source: display.getSources() )
 		{
-			sourceNameToTableDir.put( source, moBIE.getTablesDirectoryPath( (SegmentationSource) moBIE.getSource( source ) )
-			);
+			try
+			{
+				sourceNameToTableDir.put( source, moBIE.getTablesDirectoryPath( ( SegmentationSource ) moBIE.getSource( source ) )
+				);
+			} catch ( Exception e )
+			{
+				Logger.info("[WARNING] Could not store table directory for " + source );
+				sourceNameToTableDir.put( source, null );
+			}
 		}
 		display.tableViewer = new TableViewer<>( moBIE, display.tableRows, display.selectionModel, display.coloringModel, display.getName(), sourceNameToTableDir, false ).show();
 		display.selectionModel.listeners().add( display.tableViewer );

--- a/src/test/java/projects/OpenLocalCovidPlate.java
+++ b/src/test/java/projects/OpenLocalCovidPlate.java
@@ -16,8 +16,7 @@ public class OpenLocalCovidPlate
 		imageJ.ui().showUI();
 
 		try {
-			new MoBIE("/g/kreshuk/pape/Work/mobie/covid-if-project/data",
-					MoBIESettings.settings().imageDataFormat( ImageDataFormat.OmeZarr ) );
+			new MoBIE("/g/kreshuk/pape/Work/mobie/covid-if-project/data", MoBIESettings.settings().imageDataFormat( ImageDataFormat.OmeZarr ) );
 		} catch (IOException e) {
 			e.printStackTrace();
 		}

